src/components/AmplitudeProvider.js

```diff
@@ -3,62 +3,54 @@ import React from 'react';
 
 import { isValidAmplitudeInstance } from '../lib/validation';
 
-class AmplitudeProvider extends React.Component {
-  componentWillMount() {
-    const { props } = this;
+export const AmplitudeContext = React.createContext({
+  eventProperties: {},
+  instances: {},
+});
 
-    if (isValidAmplitudeInstance(props.amplitudeInstance)) {
-      if (props.apiKey) {
-        props.amplitudeInstance.init(props.apiKey);
+const AmplitudeProvider = ({ amplitudeInstance, apiKey, children, eventProperties, userId }) => {
+  const inheritedContext = React.useContext(AmplitudeContext);
+
+  const contextValue = React.useMemo(() => {
+    return {
+      eventProperties: { ...inheritedContext.eventProperties, ...eventProperties },
+      instances: {
+        ...inheritedContext.instances,
+        [amplitudeInstance._instanceName]: amplitudeInstance,
+      },
+    };
+  }, [
+    amplitudeInstance,
+    amplitudeInstance._instanceName,
+    eventProperties,
+    inheritedContext.eventProperties,
+    inheritedContext.instances,
+  ]);
+
+  React.useEffect(() => {
+    if (isValidAmplitudeInstance(amplitudeInstance)) {
+      if (apiKey) {
+        amplitudeInstance.init(apiKey);
```

Comment:
isnt apiKey required? if yes then the guarding if statement seems to be excessive

Reply:
This allows you to provide an already-initialized amplitude instance. Not sure if there's a way to identify if an amplitude instance is already-initialized to avoid re-initializing it. Something to look into…

Reply:
so generally I would make 2 possible paths here:
- require `amplitudeInstance` to be already initialized (with `apiKey` and all)
- require `apiKey` to be passed in and initialize the amplitude inside based on that

So generally allowed props would be:
```ts
{ apiKey: string, amplitudeInstance: void } | { apiKey: void, amplitudeInstance: AmplitudeInstance }
```

Current form allows for too many ways of using it (in my opinion). I understand that you might think differently or have different constraints in mind. Not a biggie anyway.